src/Adapter/MSTest.TestAdapter/Execution/TestMethodInfo.cs

```diff
@@ -942,7 +943,19 @@ private bool SetTestContext(object classInstance, TestResult result)
         {
             if (Parent.TestContextProperty != null && Parent.TestContextProperty.CanWrite)
             {
-                Parent.TestContextProperty.SetValue(classInstance, TestContext);
+                // Create a local variable to store the updated execution context
+                ExecutionContext? updatedContext = null;
+
+                // Execute the SetValue on the current execution context
+                ExecutionContextHelpers.RunOnContext(executionContext, () =>
+                {
+                    Parent.TestContextProperty.SetValue(classInstance, TestContext);
+                    // Capture the execution context after setting TestContext to preserve any AsyncLocal values set
```

Comment:
This comment seems useful, but quite generic, either remove, or add the motivation of what we are achieving here.